demos/custom-fields/field_turtle.js

```diff
@@ -0,0 +1,705 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2019 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview A field used to customize a turtle.
+ * @author bekawestberg@gmail.com (Beka Westberg)
+ */
+'use strict';
+
+// You must provide the constructor for your custom field.
+goog.provide('CustomFields.FieldTurtle');
+
+// You must require the abstract field class to inherit from.
+goog.require('Blockly.Field');
+goog.require('Blockly.utils');
+goog.require('Blockly.utils.dom');
+
+
+// Generally field's values should be optional, and have logical defaults.
+// If this is not possible (for example image fields can't have logical
+// defaults) the field should throw a clear error when a value is not provided.
+// Editable fields also generally accept validators, so we will accept a
+// validator.
+CustomFields.FieldTurtle = function(
+    opt_pattern, opt_hat, opt_turtleName, opt_validator) {
+
+  // The turtle field contains an object as its value, so we need to compile
+  // the parameters into an object.
+  var value = {};
+  value.pattern = opt_pattern;
+  value.hat = opt_hat;
+  value.turtleName = opt_turtleName;
+
+  var valid = this.doClassValidation_(value);
+  if (valid === null) {
+    // See the doClassValidation_ function for information on the
+    // cachedValidatedValue_ property.
```

Comment:
Why return null and create this cachedValidatedValue_? You'd get the same effect by setting the properties on newValue to the default when it's invalid and returning the newValue in doClassValidation_.

Reply:
You're correct that in this case it does give you the same effect, but I decided to do it this way because that's what's recommended in the [documentation](https://docs.google.com/document/d/1ONGGy0A_w24L2fDH5AqI5yQG1iyl91hEbAjMiRf3268/edit#heading=h.wcex0ucfeenc).

This is probably easier to explain with an example use case:
Say you want to create a Vector2 field that takes in text input. You would want to be able to validate the text of each number separately, and turn the text inputs red separately. To do this you would need to use the cachedValidatedValue_ pattern to save the mixed-validated state.

As you noticed the turtle field isn't a great example of this pattern, but I figured it was better to include it anyway.

Reply:
Sounds reasonable. As a followup change (separate PR) if you'd like to make it clearer how doValueInvalid_ can make use of this you could color the parts of the turtle that are invalid red (hat, shell, name) in addition to tinting the background. ;)

Reply:
Thinking about it more, styling the text of the picker or adding a warning icon next to the text would probably be clearer. That way it's easy to tell what part to change to fix the invalid selection. =)